### PR TITLE
chore(release): add automated release workflow

### PR DESCRIPTION
**Pull request by rebornplusplus**
Adds a workflow file which can be run to create a semantic tag and immediately afterwards, create a GitHub release. Any collaborator with write access should be able to trigger the workflow.

This PR also adds another workflow file, which leaves a comment after a PR has been merged. The comment reminds to trigger the aforementioned release workflow.

The tags will be automatically created using semantic versioning. However, the very first tag should be specified manually in the workflow input. The release types (patch, minor or major) are determined by the commit messages. Please see https://github.com/marketplace/actions/github-tag#bumping to learn more.

> The action will parse the new commits since the last tag using the [semantic-release](https://github.com/semantic-release/semantic-release) conventions.
> 
> semantic-release uses the commit messages to determine the type of changes in the codebase. Following formalized conventions for commit messages, semantic-release automatically determines the next [semantic version](https://semver.org) number.
> 
> By default semantic-release uses [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).
> 
> Here is an example of the release type that will be done based on a commit messages:
> 
> <table>
> <tr>
> <td> Commit message </td> <td> Release type </td>
> </tr>
> <tr>
> <td>
> 
> ```
> fix(pencil): stop graphite breaking when too much pressure applied
> ```
> 
> </td>
> <td>Patch Release</td>
> </tr>
> <tr>
> <td>
> 
> ```
> feat(pencil): add 'graphiteWidth' option
> ```
> 
> </td>
> <td>Minor Release</td>
> </tr>
> <tr>
> <td>
> 
> ```
> perf(pencil): remove graphiteWidth option
> 
> BREAKING CHANGE: The graphiteWidth option has been removed.
> The default graphite width of 10mm is always used for performance reasons.
> ```
> 
> </td>
> <td>Major Release</td>
> </tr>
> </table>

### CODE REVIEW

**Comment by SamirPS** [comment]
Can the ``create release workflow``be launched automatically after we close PR?

**Comment by rebornplusplus** [comment]
> Can the `create release workflow`be launched automatically after we close PR?

Yup, for sure, it can be. But I think we don't want to release versions after every PR merge. Some PRs may be insignificant, or sometimes irrelevant to contribute to a release.

File: .github/workflows/notify-pr-close.yml
Comment (cjdcordeiro) [inline .github/workflows/notify-pr-close.yml:16]: is there a reason why you used this action instead of something simpler like https://github.com/marketplace/actions/comment-pull-request?

File: .github/workflows/notify-pr-close.yml
Comment (cjdcordeiro) [inline .github/workflows/notify-pr-close.yml:1]: although this is cool :p I'm not too convinced of its usefulness, for a few reasons:
 1. the comment will go into a PR that may, or may NOT, have been opened by someone with write access. So in some cases, the comment just adds noise (new email) instead of adding value
 2. it works as a reminder for PRs, but what about direct pushes to `main`? I know the saying is "it is better than nothing", but it feels like a half-implementation 
 3. it isn't always relevant, i.e. we won't always want to release, so why send a reminder?

If you're really keen on keeping this and giving it a try, I won't oppose to it, but my gut feeling is that it will become a nuisance that we all just ignore cause we (writers) know what the process is to get a new release.

File: .github/workflows/notify-pr-close.yml
Comment (rebornplusplus) [reply .github/workflows/notify-pr-close.yml:16]: The only reason for me was that [github-script](https://github.com/marketplace/actions/github-script) is marked verified on their marketplace, while the other one is not.

> Verified creator
> GitHub has verified that this action was created by https://github.com/actions.
> 
> [Learn more about verified Actions](https://docs.github.com/actions/automating-your-workflow-with-github-actions/using-github-marketplace-actions).

But github-script is a bit more generic than [Comment Pull Request](https://github.com/marketplace/actions/comment-pull-request). Let me know if you think I should use the latter.

File: .github/workflows/release.yml
Comment (cjdcordeiro) [inline .github/workflows/release.yml:1]:
```suggestion
name: Tag and Release
```

File: .github/workflows/release.yml
Comment (cjdcordeiro) [inline .github/workflows/release.yml:22]: I think this should be conditional. I.e., you want an automatic bump whenever there is no inputs.tag, right? Like, if the previous release was v1.2.3, and I trigger the workflow without any inputs, then I'm expecting an automatic new version v1.2.4. Otoh, if there is `inputs.tag`, this shall be `false`

File: .github/workflows/release.yml
Comment (cjdcordeiro) [inline .github/workflows/release.yml:23]: you can just use `${{ inputs.tag }}`

File: .github/workflows/release.yml
Comment (cjdcordeiro) [inline .github/workflows/release.yml:17]: what happens if the tag already exists?

File: .github/workflows/release.yml
Comment (cjdcordeiro) [inline .github/workflows/release.yml:7]: is `required: true` the default?

File: .github/workflows/release.yml
Comment (cjdcordeiro) [inline .github/workflows/release.yml:30]: You should also attach the relevant scripts/binaries to the release

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
Cool, thanks! Left a few comments. Do you have an example of this workflow's results?

File: .github/workflows/notify-pr-close.yml
Comment (rebornplusplus) [reply .github/workflows/notify-pr-close.yml:1]: Ah okay, good points. I will just ditch it for now.

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml:17]: Right, I need to test this use case. Good catch, forgot to test this one.

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml:7]: I think `required: false` is the default. But in any case, I will make sure to specify it.

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml:30]: Right, you mean the new/changed files introduced in this release?

File: .github/workflows/notify-pr-close.yml
Comment (rebornplusplus) [reply .github/workflows/notify-pr-close.yml]: Removed in https://github.com/canonical/rocks-toolbox/pull/2/commits/b284546e3b81a8032bd2d5e99bcc671ae914b6de.

File: .github/workflows/notify-pr-close.yml
Comment (rebornplusplus) [reply .github/workflows/notify-pr-close.yml]: Irrelevant now since this is removed in https://github.com/canonical/rocks-toolbox/pull/2/commits/b284546e3b81a8032bd2d5e99bcc671ae914b6de.

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml]: Maybe I misunderstood this a bit previously. I added a condition where `default_bump: false` if there is a non-empty `inputs.tag`. Otherwise, I kept the default to `patch`. Thanks!

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml]: Nice, thanks.

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml:18]: The tag creation will be skipped. Per the changed release workflow in https://github.com/canonical/rocks-toolbox/pull/2/commits/b1d4989eebc6a60dc2056e0999cae459d43b0065, the release will be skipped too if the release for that tag exists. However, if a tag exists and the release for that tag doesn't, a new release will be created.

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml]: I am a bit confused about this one. Which files do I attach?

File: .github/workflows/release.yml
Comment (cjdcordeiro) [reply .github/workflows/release.yml]: Thanks. Won't the `custom_tag` interfere with the automatic bump? Cause if `inputs.tag == ''` then `custom_tag: ''`...is that a problem?

File: .github/workflows/release.yml
Comment (cjdcordeiro) [reply .github/workflows/release.yml]: I think all, even the ones that haven't changed. If we look at this repo as a collection of utilities, on every release, all utilities should be attached to the GH release.

So for now, just your bash script.

**Review by cjdcordeiro — COMMENTED** [review]
thanks for the example and diagram. It's LGTM...just need to polish the last two comments and that should be it

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml]: No. If `custom_tag: ''`, it is ignored and the automatic bump takes over.

File: .github/workflows/release.yml
Comment (rebornplusplus) [reply .github/workflows/release.yml]: Added in https://github.com/canonical/rocks-toolbox/pull/2/commits/34f15ed72aa500aae3d6a1ce985f0ad1cb0d9f93.
Demo for the attached files: https://github.com/rebornplusplus/rocks-toolbox/releases/tag/v1.0.9.

**Review by cjdcordeiro — APPROVED** [review]

**Comment by cjdcordeiro** [comment]
Great, thanks. merging

**Pull request merged**
